src/engine/controls/bpmcontrol.cpp

```diff
@@ -725,6 +726,162 @@ double BpmControl::getNearestPositionInPhase(
     return dNewPlaypos;
 }
 
+double BpmControl::getBeatMatchPosition(
+        double dThisPosition, bool respectLoops, bool playing) {
+    // Without a beatgrid, we don't know the phase offset.
+    if (!m_pBeats) {
+        return dThisPosition;
+    }
+    // Master buffer is always in sync!
+    if (getSyncMode() == SYNC_MASTER) {
+        return dThisPosition;
+    }
+
+    // Get the current position of this deck.
+    double dThisPrevBeat = m_pPrevBeat->get();
+    double dThisNextBeat = m_pNextBeat->get();
+    double dThisBeatLength;
+    if (dThisPosition > dThisNextBeat || dThisPosition < dThisPrevBeat) {
+        qDebug() << "BpmControl::getNearestPositionInPhase out of date"
+                << dThisPosition << dThisNextBeat << dThisPrevBeat;
+        // This happens if dThisPosition is the target position of a requested
+        // seek command
+        if (!getBeatContext(m_pBeats, dThisPosition,
+                            &dThisPrevBeat, &dThisNextBeat,
+                            &dThisBeatLength, nullptr)) {
+            return dThisPosition;
+        }
+    } else {
+        if (!getBeatContextNoLookup(dThisPosition,
+                                    dThisPrevBeat, dThisNextBeat,
+                                    &dThisBeatLength, nullptr)) {
+            return dThisPosition;
+        }
+    }
+
+    double dOtherPrevBeat;
+    double dOtherNextBeat;
+    double dOtherBeatLength;
+    double dOtherBeatFraction;
+    // If not, we have to figure it out
+    EngineBuffer* pOtherEngineBuffer = pickSyncTarget();
+    if (pOtherEngineBuffer == nullptr) {
+        if (playing) {
+            // Sync to itselfe if we are already playing
+            pOtherEngineBuffer = getEngineBuffer();
+        } else {
+            return dThisPosition;
+        }
+    }
+
+    if (playing) {
+        // "this" track is playing, or just starting
+        // only match phase if the sync target is playing as well
+        if (pOtherEngineBuffer->getSpeed() == 0.0) {
+            return dThisPosition;
+        }
+    }
+
+    TrackPointer otherTrack = pOtherEngineBuffer->getLoadedTrack();
+    BeatsPointer otherBeats = otherTrack ? otherTrack->getBeats() : BeatsPointer();
+
+    // If either track does not have beats, then we can't adjust the phase.
+    if (!otherBeats) {
+        return dThisPosition;
+    }
+
+    double dOtherPosition = pOtherEngineBuffer->getExactPlayPos();
+
+    if (!BpmControl::getBeatContext(otherBeats, dOtherPosition,
```

Comment:
I don't think this code correctly handles half/double bpm sync situations.  In trying to merge this with my one mastersync-onemaster code, it picks the wrong beats to sync against.